src/main/kotlin/ecommerce/service/CartItemService.kt

```diff
@@ -0,0 +1,113 @@
+import ecommerce.repository.ProductOptionRepository
+import jakarta.transaction.Transactional
+import org.springframework.stereotype.Service
+import java.time.LocalDateTime
+import kotlin.jvm.optionals.getOrNull
+
+@Service
+class CartItemService(
+    private val cartRepository: CartRepository,
+    private val cartItemRepository: CartItemRepository,
+    private val productOptionRepository: ProductOptionRepository,
+) {
+    @Transactional
+    fun saveCartItem(
+        request: AddToCartRequest,
+        cartId: Long,
+        cartItemId: Long? = null,
+    ): CartItem {
+        val cart = findCartById(cartId)
+        val productOption = findProductOptionById(request.productOptionId)
+        val existingCartItem = findExistingCartItem(cartItemId, cart, productOption)
+
+        return if (existingCartItem != null) {
+            updateExistingCartItem(existingCartItem, request, productOption, cart, cartItemId != null)
+        } else {
+            createNewCartItem(cart, productOption, request)
+        }
+    }
+
+    private fun findCartById(cartId: Long) =
+        cartRepository.findById(cartId).getOrNull()
+            ?: throw NotFoundException("Cart not found")
+
+    private fun findProductOptionById(productOptionId: Long) =
+        productOptionRepository.findById(productOptionId).getOrNull()
+            ?: throw NotFoundException("Product option not found")
+
+    private fun findExistingCartItem(
+        cartItemId: Long?,
+        cart: ecommerce.model.Cart,
+        productOption: ecommerce.model.ProductOption,
+    ) = cartItemId?.let { cartItemRepository.findById(it).getOrNull() }
+        ?: cartItemRepository.findByCartAndProductOption(cart, productOption)
+
+    private fun updateExistingCartItem(
+        existingCartItem: CartItem,
+        request: AddToCartRequest,
+        productOption: ecommerce.model.ProductOption,
+        cart: ecommerce.model.Cart,
+        isDirectUpdate: Boolean,
+    ): CartItem {
+        if (isDirectUpdate) {
+            productOption.checkQuantityIncrement(request.newProductOptionQuantity)
+            productOption.updateQuantity(request.newProductOptionQuantity)
+            productOptionRepository.save(productOption)
+            cart.addQuantity(request.newProductOptionQuantity)
+        }
+
+        val newQuantity =
+            if (isDirectUpdate) {
+                request.newProductOptionQuantity
+            } else {
+                existingCartItem.quantity + request.newProductOptionQuantity
+            }
+
+        CartItem.validateQuantity(newQuantity)
+        existingCartItem.modify(null, null, newQuantity, LocalDateTime.now())
+        return existingCartItem
+    }
+
+    private fun createNewCartItem(
+        cart: ecommerce.model.Cart,
+        productOption: ecommerce.model.ProductOption,
+        request: AddToCartRequest,
+    ): CartItem {
+        CartItem.validateQuantity(request.newProductOptionQuantity)
+
+        return cartItemRepository.save(
```

Comment:
> I thought it could be also easy to understand and handle them by locating related methods in one file: CartItemService. But I understand that your way is easier.

I see.... From what I saw in CartItemServiceTest, the procedure to run the test seems much longer compared to when only using ProductOption. What are your thoughts on that?